path_role.go

```diff
@@ -0,0 +1,433 @@
+			},
+		},
+
+		ExistenceCheck: b.pathRoleExistenceCheck,
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.CreateOperation: b.pathRoleCreate,
+			logical.UpdateOperation: b.pathRoleUpdate,
+			logical.ReadOperation:   b.pathRoleRead,
+			logical.DeleteOperation: b.pathRoleDelete,
+		},
+
+		HelpSynopsis:    pathRoleSyn,
+		HelpDescription: pathRoleDesc,
+	}
+}
+
+func pathListRoles(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "role/?",
+
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.ListOperation: b.pathRoleList,
+		},
+
+		HelpSynopsis:    pathListRolesHelpSyn,
+		HelpDescription: pathListRolesHelpDesc,
+	}
+}
+
+// Establishes dichotomy of request operation between CreateOperation and UpdateOperation.
+// Returning 'true' forces an UpdateOperation, CreateOperation otherwise.
+func (b *backend) pathRoleExistenceCheck(ctx context.Context, req *logical.Request, data *framework.FieldData) (bool, error) {
+	entry, err := b.lockedOCIRole(ctx, req.Storage, strings.ToLower(data.Get("role").(string)))
+	if err != nil {
+		return false, err
+	}
+	return entry != nil, nil
+}
+
+// lockedOCIRole returns the properties set on the given role. This method
+// acquires the read lock before reading the role from the storage.
+func (b *backend) lockedOCIRole(ctx context.Context, s logical.Storage, roleName string) (*OCIRoleEntry, error) {
+	if roleName == "" {
+		return nil, fmt.Errorf("missing role name")
+	}
+
+	b.roleMutex.RLock()
+	roleEntry, err := b.nonLockedOCIRole(ctx, s, roleName)
+	// we manually unlock rather than defer the unlock because we might need to grab
+	// a read/write lock in the upgrade path
+	b.roleMutex.RUnlock()
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return nil, nil
+	}
+	return roleEntry, nil
+}
+
+// lockedSetOCIRole creates or updates a role in the storage. This method
+// acquires the write lock before creating or updating the role at the storage.
+func (b *backend) lockedSetOCIRole(ctx context.Context, s logical.Storage, roleName string, roleEntry *OCIRoleEntry) error {
+	if roleName == "" {
+		return fmt.Errorf("missing role name")
+	}
+
+	if roleEntry == nil {
+		return fmt.Errorf("nil role entry")
+	}
+
+	b.roleMutex.Lock()
+	defer b.roleMutex.Unlock()
+
+	return b.nonLockedSetOCIRole(ctx, s, roleName, roleEntry)
+}
+
+// nonLockedSetOCIRole creates or updates a role in the storage. This method
+// does not acquire the write lock before reading the role from the storage. If
+// locking is desired, use lockedSetOCIRole instead.
+func (b *backend) nonLockedSetOCIRole(ctx context.Context, s logical.Storage, roleName string,
+	roleEntry *OCIRoleEntry) error {
+	if roleName == "" {
+		return fmt.Errorf("missing role name")
+	}
+
+	if roleEntry == nil {
+		return fmt.Errorf("nil role entry")
+	}
+
+	entry, err := logical.StorageEntryJSON("role/"+strings.ToLower(roleName), roleEntry)
```

Comment:
Set the role name parameter in the path definition to `framework.TypeLowerCaseString` and then you won't have to worry about `ToLower`s... the framework will lowercase the parameter before your code gets it.

Reply:
Done, thanks!